sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/exemplar/FixedSizeExemplarReservoir.java

```diff
@@ -0,0 +1,182 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.sdk.metrics.exemplar;
+
+import io.opentelemetry.api.common.AttributeKey;
+import io.opentelemetry.api.common.Attributes;
+import io.opentelemetry.api.common.AttributesBuilder;
+import io.opentelemetry.api.internal.GuardedBy;
+import io.opentelemetry.api.trace.Span;
+import io.opentelemetry.context.Context;
+import io.opentelemetry.sdk.common.Clock;
+import io.opentelemetry.sdk.metrics.data.DoubleExemplar;
+import io.opentelemetry.sdk.metrics.data.Exemplar;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.concurrent.atomic.LongAdder;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import javax.annotation.Nullable;
+
+/**
+ * A Reservoir sampler with fixed size that stores the given number of exemplars.
+ *
+ * <p>This implementation uses a un-unweighted/naive algorithm for sampler where the probability of
+ * sampling decrease as the number of observations continue. The collectAndReset method resets the
+ * count of observations, making the probability of sampling effectively 1.0.
+ *
+ * <p>Additionally this implementation ONLY exports double valued exemplars.
+ */
+public class FixedSizeExemplarReservoir implements ExemplarReservoir {
+  private final Clock clock;
+  private final ReservoirCell[] storage;
+  private final LongAdder numMeasurements = new LongAdder();
+
+  /**
+   * Instantiates an exemplar reservoir of fixed size.
+   *
+   * @param clock The clock to use when annotating measurements with time.
+   * @param size The number of exemplars to preserve.
+   */
+  public FixedSizeExemplarReservoir(Clock clock, int size) {
+    this.clock = clock;
+    this.storage = new ReservoirCell[size];
+    for (int i = 0; i < size; ++i) {
+      this.storage[i] = new ReservoirCell();
+    }
+  }
+
+  protected int maxSize() {
+    return storage.length;
+  }
+  /**
+   * Determines the sample bucket for a given measurement.
+   *
+   * @return The bucket to sample into or -1 for no sampling.
+   */
+  protected int bucketFor(double value, Attributes attributes, Context context) {
+    // TODO: hijackable random for testing.
+    long index = ThreadLocalRandom.current().nextLong(numMeasurements.sum() + 1);
```

Comment:
Just FYI, TLR is super flaky and unreliable on Android (depending on API version). It is always initialized with the same seed, making it terrible for anything you actually want to be random.

Reply:
Agreed.  WDYT of adding something like this - https://github.com/census-instrumentation/opencensus-java/blob/master/impl_core/src/main/java/io/opencensus/implcore/trace/internal/RandomHandler.java

Specifically, I think there's a lot of reasons NOT to use random (or TLR) directly

Reply:
right now, I'm sure you've seen, we have a quick-and-dirty Android detector to decide which IdGenerator to use. I could imagine we abstract that to include all random's that are needed in the SDK.

Reply:
separate PR, or include in this one?

Reply:
If you want a separate PR, please create an issue to track it so we don't forget about it.

Reply:
Added it in this PR, but definitely needs a look over from you and @anuraaga